### PR TITLE
chore: Update Emotion to v11

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request
Update Emotion packages to v11.

The main changes needed were:
1. Rename the packages per https://emotion.sh/docs/emotion-11. (E.g. `@emotion/core` -> `@emotion/react`)
2. Update to the new way of defining themes in Typescript (via a module declaration on `@emotion/react`, and then import `styled` directly from `@emotion/styled`)
3. Manually resolve emotion 11 in storybook. Storybook internally tries to resolve emotion 10 still, which breaks things.
4. Add some return types to appease typescript
5. Remove manual `PropTypes` in some cases. Emotion 11 has better typescript support, so if the `PropTypes` did not exactly match the typescript types, it would fail. So I removed manual `PropTypes` declarations in two places. (They now rely on the Typescript prop types)

### Testing instructions
1. TeamCity should pass, other than some style issues related to use before define (fixed in #55806)
2. `yarn` should succeed locally
3. Storybook should work. In particular, try `composite-checkout:storybook:start` and verify that the components show up in the browser.
4. yarn start/build should work.
5. Smoke test calypso.live

### CODE REVIEW

**Review by sirbrillig — APPROVED** [review]
On behalf of Payments: the code looks good and checkout still appears to work and look correct. Thank you!

File: packages/components/.storybook/main.js
Comment (scinos) [inline packages/components/.storybook/main.js]: `path.join( process.cwd(), 'node_modules/@emotion/styled' )` seems a bit weak: if the process to launch storybook is run from a different directory, it will generate the wrong path. Plus if `@emotion` get hoisted differently it may point to the wrong version.

What about using require.resolve to find `@emotion` path?

**Comment by scinos** [comment]
Shouldn't we need to declare a dependency on `@emotion` v11 in some package.json?

File: packages/composite-checkout/src/components/checkout-error-boundary.tsx
Comment (ciampo) [inline packages/composite-checkout/src/components/checkout-error-boundary.tsx]: Maybe a better type for `children` would be [`React.ReactNode`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L491)?

File: packages/composite-checkout/src/components/checkout-error-boundary.tsx
Comment (ciampo) [inline packages/composite-checkout/src/components/checkout-error-boundary.tsx]: Maybe a better return type for `render` would be [`React.ReactNode`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L491)?

File: packages/composite-checkout/src/components/checkout-order-summary.tsx
Comment (ciampo) [inline packages/composite-checkout/src/components/checkout-order-summary.tsx:17]: Maybe a better type would be [`React.FunctionComponent`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L543) ?

(this comment would apply to every other instance of React's functional component types with this pattern)

File: packages/composite-checkout/src/components/spinner.tsx
Comment (ciampo) [inline packages/composite-checkout/src/components/spinner.tsx]: Just another example of how we could use `React. FunctionComponent` (also aliased to `React.FC`):

```typescript
const Spinner: React.FC< { className?: string } > = ( { className } ) => (
	<SpinnerWrapper className={ className } />
);
export default Spinner;
```

File: packages/composite-checkout/src/custom.d.ts
Comment (ciampo) [inline packages/composite-checkout/src/custom.d.ts:7]: What is the reason for adding the `Theme` interface?

Also, isn't the `@emotion/react` module already been "augmented" in `packages/composite-checkout/src/types.ts` ?

File: package.json
Comment (ciampo) [inline package.json:172]: Was this dependency unneeded, or is this change caused by the migration to Emotion 11?

Same question would apply for the `jest-emotion` dependency deletion in `packages/calypso-build/package.json`

**Review by ciampo — COMMENTED** [review]
Hey Noah, great job! Thank you for looking into the migration to Emotion 11.

Although I'm not directly involved, I've had a look and left a few inline comments :)

File: packages/composite-checkout/src/components/checkout-order-summary.tsx
Comment (sirbrillig) [reply packages/composite-checkout/src/components/checkout-order-summary.tsx:17]: IMO using `JSX.Element` is nicer because you can use the `export function` syntax and don't need to do `const` with an arrow function and then `export const`.

File: packages/composite-checkout/src/custom.d.ts
Comment (noahtallen) [reply packages/composite-checkout/src/custom.d.ts:7]: > Also, isn't the @emotion/react module already been "augmented" in packages/composite-checkout/src/types.ts ?

Great question! That augmentation is actually the old way of defining themes. The new recommended way is to augment the theme type like here, and then we can import directly from `@emotion/styled` instead of importing from a custom styled file. More here: https://emotion.sh/docs/emotion-11#theme-type

The `styled` files should be deleted in this PR as well

File: package.json
Comment (noahtallen) [reply package.json:172]: I made this change when working through all the dependency issues raised by yarn. I suspect this can be restored. `jest-emotion` didn't appear to be used by calypso-build, which is why I removed it. (We only use it in `client/`, so it's added to that package.json)

File: packages/components/.storybook/main.js
Comment (noahtallen) [reply packages/components/.storybook/main.js]: I can look into that :+1:

**Comment by noahtallen** [comment]
> Shouldn't we need to declare a dependency on @emotion v11 in some package.json?

According to https://emotion.sh/docs/install, we just need `@emotion/react`. This used to be named `@emotion/core`

**Comment by noahtallen** [comment]
> I've had a look and left a few inline comments

Thanks for all the TS tips!! I'll defer to you on those, as I don't have much experience with best practices :)

File: packages/composite-checkout/src/components/spinner.tsx
Comment (sirbrillig) [reply packages/composite-checkout/src/components/spinner.tsx]: Definitely true, but (as above) for now I'd prefer to keep this code using `export default function`.

Comment (noahtallen) [reply packages/composite-checkout/src/components/spinner.tsx]: I'll keep it as is then

File: package.json
Comment (noahtallen) [reply package.json:172]: Oh, the other reason is that `@types/jest` needs to be at a lower version in one of the packages to meet peer dependency requirements.

File: packages/composite-checkout/src/custom.d.ts
Comment (noahtallen) [reply packages/composite-checkout/src/custom.d.ts:7]: I think the concept is that `@emotion/` infers the correct `Theme` type internally instead of using a custom type everywhere we use `styled`

File: apps/editing-toolkit/tsconfig.json
Comment (scinos) [inline apps/editing-toolkit/tsconfig.json:33]: Not for this PR, but I suspect we can remove `@emotion/react` from here. Probably even remove `types` entirely.

File: client/package.json
Comment (scinos) [inline client/package.json:248]: Does this need to match `jest` version?

Comment (noahtallen) [reply client/package.json:248]: It was previously on v25 (with jest at v27), so it wasn't causing issues. I had to set it to 26 because `@emotion/jest` sets `@types/jest@26` as a peer dependency.

Comment (noahtallen) [reply client/package.json:248]: is there a better way to handle that?

Comment (scinos) [reply client/package.json:248]: Not really, if the authors of `@emotion/jest` really want v26, I'd say we go with that unless we have a very strong reason to override their decision.

That being said, I don't understand why they would want a specific version for _types_, as we won't be type checking `@emotion/jest` but our own code 🤷🏼

Comment (noahtallen) [reply client/package.json:248]: Yeah, I have no idea: https://github.com/emotion-js/emotion/blob/main/packages/jest/package.json#L27

File: packages/composite-checkout/src/types.ts
Comment (scinos) [inline packages/composite-checkout/src/types.ts:6]: For anyone trying to understand what is going on here: https://emotion.sh/docs/emotion-11#theme-type

Comment (noahtallen) [reply packages/composite-checkout/src/types.ts:6]: yup. once you declare the theme on the module here, `props.theme` automatically gets the correct types in this kind of example:

```js
import styled from '@emotion/styled';

const Thing = styled.div`
  color: ${ ( props ) => props.theme.colors.highlight }; // `theme` type is inferred from our module declaration
`
```

File: yarn.lock
Comment (scinos) [inline yarn.lock:3350]: Looks like we will be adding `@emotion/styled@11.3.0` but `@emotion/styled@10.0.27` is still in the system. This applies to other `@emotion/*` packages in this PR too. 

This is not a bad thing _per se_, but it might indicate we still have Emotion v10 around. It might be worth double checking.

**Comment by scinos** [comment]
I reviewed this from the package management point of view and everything makes sense to me 👍🏼

**Review by scinos — APPROVED** [review]

File: yarn.lock
Comment (noahtallen) [reply yarn.lock:3350]: I'll double check, but I think that comes from wordpress. The wp packages still require emotion 10. After our react 17 + wp update lands, those should also support emotion 11

Comment (noahtallen) [reply yarn.lock:3350]: Yep, v10 comes from `@storybook/theming` and `@wordpress/components`, which are both expected

File: packages/components/.storybook/main.js
Comment (noahtallen) [reply packages/components/.storybook/main.js]: yep, this worked!

**Comment by ciampo** [comment]
> Manually resolve emotion 11 in storybook. Storybook internally tries to resolve emotion 10 still, which breaks things.

It looks like Storybook won't support Emotion 11 for a while, and that potentially may move away from Emotion at some point: https://github.com/storybookjs/storybook/issues/13145#issuecomment-908876098

**Comment by rrsai** [comment]
After much debugging of the clash between versions of emotion used in storybook and in app, and for those who come after me, I found that storybook has now provided a feature flag to avoid clashing with later versions of emotion: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#emotion11-quasi-compatibility
I tested it and it works well to prevent emotion version-out-of-sync problems in storybook, and is a 3 line change.

**Comment by noahtallen** [comment]
> I found that storybook has now provided a feature flag to avoid clashing with later versions of emotion

Nice find! Edit: Since it's only available in newer versions, I'm looking at doing that in #58880

**Pull request merged**
